File: 2025-1/sesion-de-clase-24-net/CuentaBancaria.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace sesion_de_clase_24_net
{
    class CuentaBancaria
    {
        private double saldo;
        private readonly Object lockObj = new Object();

        public CuentaBancaria(double saldo) { 
            this.saldo = saldo;
        }

        public void Retirar(string nombreCliente, double monto) {
            lock (lockObj) {
                while (saldo < monto)
                {
                    Console.WriteLine(nombreCliente + " quiere retirar " + monto + " pero no hay suficiente, esperando...");
                    Monitor.Wait(lockObj);
                }

                Console.WriteLine(nombreCliente + " va a retirar " + monto);
                this.saldo -= monto;
                Console.WriteLine(nombreCliente + " retiro " + monto + " queda " + this.saldo);
            }
        }

        public void Depositar(string nombreCliente, double monto) {
            lock (lockObj) {
                Console.WriteLine(nombreCliente + " deposito " + monto);

                this.saldo += monto;

                Console.WriteLine("Saldo luego del deposito: " + this.saldo);

                Monitor.PulseAll(lockObj);
            }
        }

    }
}

File: 2025-2/sesion-de-clase-24/java/src/main/java/CuentaBancaria.java


/**
 *
 * @author eric
 */
public class CuentaBancaria {
    private double saldo = 500.00;
    
    public synchronized void retirar(String nombreCliente, double monto) {
        while (this.saldo < monto) {
            System.out.println(nombreCliente + " quiere retirar " + monto + " pero no hay sificiente, esperando...");
            try {
                wait();
            }
            catch (InterruptedException ex) {
                Thread.currentThread().interrupt();
            }
        }
               
        System.out.println("El cliente " + nombreCliente + " va a retirar " + monto);

        this.saldo -= monto;
        System.out.println(nombreCliente + " completo el retiro, queda " + this.saldo);
    }
    
    public synchronized void depositar(String nombreCliente, double monto) {
        System.out.println("El cliente " + nombreCliente + " deposito " + monto);
        saldo += monto;
        System.out.println("Saldo despues del deposito: " + this.saldo);
        notifyAll();
    }
}
